### PR TITLE
introduce flash_area_write_data

### PR DESCRIPTION
**Pull request by TychoVrahe**
Introducing flash_area_write_data function, which aims to simplify writing arbitrarily large data to flash.

### CODE REVIEW

File: core/embed/firmware/bl_check.c [hunk 1]
```diff
@@ -217,27 +217,20 @@ void check_and_replace_bootloader(void) {
...
+    offset += size_padded;
     decomp.dest = (uint8_t *)decomp_out;
   } while (uzlib_uncompress(&decomp) >= 0);
 
-  uint32_t d[FLASH_BURST_WORDS] = {0};
-  // fill the rest of the bootloader area with 0x00
-  while (offset < bl_len) {
-    ensure(flash_area_write_burst(&BOOTLOADER_AREA, offset, d), NULL);
-    offset += FLASH_BURST_WORDS * sizeof(uint32_t);
+  if (bl_len - offset > 0) {
```
Comment (cepetr) [inline core/embed/firmware/bl_check.c:229]: nothing wrong on it, just consider a bit clearer condition: `if (offset < bl_len) {`

**Review by cepetr — CHANGES_REQUESTED** [review]
I like this change very much.  I just added few comments.

File: core/embed/boardloader/main.c [hunk 2]
```diff
@@ -227,11 +227,9 @@ static secbool copy_sdcard(void) {
   // copy bootloader from SD card to Flash
   term_printf("copying new bootloader from SD card\n\n");
 
-  for (int j = 0; j < (IMAGE_HEADER_SIZE + codelen) / FLASH_BURST_SIZE; j++) {
-    ensure(flash_area_write_burst(&BOOTLOADER_AREA, j * FLASH_BURST_SIZE,
-                                  &sdcard_buf[j * FLASH_BURST_WORDS]),
-           NULL);
-  }
+  ensure(flash_area_write_data(&BOOTLOADER_AREA, codelen, sdcard_buf,
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/boardloader/main.c]: the whole thing was wildly incorrect: https://github.com/trezor/trezor-firmware/commit/bbe1951e8ea0aef7b5a90e41ca6e5710a3553429

File: storage/flash_area.c [hunk 3]
```diff
@@ -145,6 +149,119 @@ secbool flash_area_write_block(const flash_area_t *area, uint32_t offset,
...
+  if (offset % FLASH_BLOCK_SIZE) {
+    return secfalse;
+  }
+  if (total_size % FLASH_BLOCK_SIZE) {
+    return secfalse;
+  }
+  if (data_size > total_size) {
+    return secfalse;
+  }
+  if (total_size > flash_area_get_size(area)) {
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply storage/flash_area.c]: yes fixed https://github.com/trezor/trezor-firmware/commit/bbe1951e8ea0aef7b5a90e41ca6e5710a3553429

File: storage/flash_area.c [hunk 4]
```diff
@@ -145,6 +149,119 @@ secbool flash_area_write_block(const flash_area_t *area, uint32_t offset,
...
+
+    if (flash_area_write_block(area, offset, data_qw) != sectrue) {
+      return secfalse;
+    }
+    offset += FLASH_BLOCK_SIZE;
+    data32 += FLASH_BLOCK_WORDS;
+    total_size -= FLASH_BLOCK_SIZE;
+  }
+  return sectrue;
+}
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply storage/flash_area.c:234]: yeah love it, its much better: https://github.com/trezor/trezor-firmware/commit/bbe1951e8ea0aef7b5a90e41ca6e5710a3553429

File: core/embed/firmware/bl_check.c (hunk 1 above)
Comment (TychoVrahe) [reply core/embed/firmware/bl_check.c]: fixed https://github.com/trezor/trezor-firmware/commit/bbe1951e8ea0aef7b5a90e41ca6e5710a3553429

**Review by cepetr — APPROVED** [review]

**Pull request merged**